File: LP/projectEuler/lychrel.py
# Função que retorna true se um número for palíndromo
def palindromo(num):
        rev = str(num)[::-1]
        if str(num) == rev: return True
        else: return False

# Retorna falso caso encontre um palíndromo ao somar um número
# a seu reverso. Esse procedimento é feito 50 vezes. Se não
# for encontrado um palíndromo dentro desse limite de iterações,
# então o número é de Lychrel e retorna-se verdadeiro
def isLychrel (num):
        atual = num
        for x in range(50):
                atual += int(str(atual)[::-1])
                if palindromo(atual): return False
        return True

# Cria um contador para incrementar a cada número de Lychrel
contador = 0

# Para esse problema, foi solicitado analisar números de 0 a 10000
for i in range(1, 10000):
        lych_num = isLychrel(i)
        if (lych_num): contador += 1

print("Quantidade: " + str(contador))

File: LP/projectEuler/lychrel.pl
/*função para converter um inteiro em dígito*/
digit(N,X)
	:- digit(N,[],X), !.
digit(0,L,L).
digit(N,L,X)
	:- N1 is N//10, N2 is N rem 10, digit(N1,[N2|L],X).

/*função para converter um dígito em inteiro*/
integer([],0).
integer([A|As],X)
	:- integer(A,As,X), !.
integer(A,[],A).
integer(A,[B|Bs],X)
	:- C is 10*A+B, integer(C,Bs,X).

/*retorna 1 em caso de um número ser de lychrel*/
/*este procedimento é feito para cada valor entre 1 e 9999*/
num(1)
	:- between(1,9999,N), lychrel(N).

lychrel(N)
	:- lychrel(N,0).

/*realiza o procedimento 50 vezes*/
lychrel(_,V)
	:- V >= 50, !.
	
/*retorna true para um número não palíndromo*/
lychrel(N,V)
	:- rev_sum(N,R), not(is_palin(R)), succ(V,V1), lychrel(R,V1).

/*função append tirada das notas de aula de prolog*/
myAppend([], L, L).
myAppend([Head|TailA], B, [Head|TailC])
	:- myAppend(TailA, B, TailC).

/*função para reverso de um número tirada das notas de aula de prolog*/
myReverse([], []).
myReverse([H|T], X)
	:- myReverse(T, XAux), myAppend(XAux, [H], X).

/*função para somar um número ao reverso dele utilizando as funções de conversão inteiro-dígito*/
rev_sum(N,X)
	:- digit(N,L), myReverse(L,L1), integer(L1,N1), X is N+N1.

/*retorna true caso o valor seja palíndromo*/
is_palin(N)
	:- digit(N,L), myReverse(L,L).

result(X)
	:- aggregate(sum(N), num(N), X).
